### PR TITLE
[SPARK-30181][SQL] Only add string or integral type column to metastore partition filter

### PR DESCRIPTION
**Pull request by yujhe-li**
### What changes were proposed in this pull request?

The following SQL throws runtime exception since Spark will cast `dt` as string and add the comparison to partition filter. (the hive metastore filter: 'dt >= "2019-12-01 00:00:00"')

The reason is that Hive metastore client only support filtering on partition keys of string or integral type. 
https://github.com/apache/hive/blob/release-1.2.1/metastore/src/java/org/apache/hadoop/hive/metastore/parser/ExpressionTree.java#L433

```scala
val df = Seq(
    (1, java.sql.Timestamp.valueOf("2019-12-01 00:00:00"), 1), 
    (2, java.sql.Timestamp.valueOf("2019-12-01 01:00:00"), 1)
  ).toDF("id", "dt", "value")
df.write.partitionBy("dt").mode("overwrite").saveAsTable("timestamp_part")

spark.sql("select * from timestamp_part where dt >= '2019-12-01 00:00:00'").explain(true)
Caught Hive MetaException attempting to get partition metadata by filter from Hive. You can set the Spark configuration setting spark.sql.hive.manageFilesourcePartitions to false to work around this problem, however this will result in degraded performance. Please report a bug: https://issues.apache.org/jira/browse/SPARK
java.lang.RuntimeException: Caught Hive MetaException attempting to get partition metadata by filter from Hive. You can set the Spark configuration setting spark.sql.hive.manageFilesourcePartitions to false to work around this problem, however this will result in degraded performance. Please report a bug: https://issues.apache.org/jira/browse/SPARK
  at org.apache.spark.sql.hive.client.Shim_v0_13.getPartitionsByFilter(HiveShim.scala:774)
  at org.apache.spark.sql.hive.client.HiveClientImpl$$anonfun$getPartitionsByFilter$1.apply(HiveClientImpl.scala:679)
  at org.apache.spark.sql.hive.client.HiveClientImpl$$anonfun$getPartitionsByFilter$1.apply(HiveClientImpl.scala:677)
  at org.apache.spark.sql.hive.client.HiveClientImpl$$anonfun$withHiveState$1.apply(HiveClientImpl.scala:275)
  at org.apache.spark.sql.hive.client.HiveClientImpl.liftedTree1$1(HiveClientImpl.scala:213)
...
Caused by: MetaException(message:Filtering is supported only on partition keys of type string)
  at org.apache.hadoop.hive.metastore.parser.ExpressionTree$FilterBuilder.setError(ExpressionTree.java:185)
  at org.apache.hadoop.hive.metastore.parser.ExpressionTree$LeafNode.getJdoFilterPushdownParam(ExpressionTree.java:440)
  at org.apache.hadoop.hive.metastore.parser.ExpressionTree$LeafNode.generateJDOFilterOverPartitions(ExpressionTree.java:357)
```

This PR avoid adding partition keys to filter which its data type is not equals string or integral.

### Does this PR introduce any user-facing change?
Yes, the above SQL will go to get all partitions from metastore without runtime exception.

### How was this patch tested?
New test added.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 1]
```diff
@@ -676,7 +676,9 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
     object ExtractAttribute {
       def unapply(expr: Expression): Option[Attribute] = {
         expr match {
-          case attr: Attribute => Some(attr)
+          case attr: Attribute
+              if attr.dataType == StringType || IntegralType.acceptsType(attr.dataType) =>
```
Comment (seanli-rallyhealth) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: you may define a private function for this.

Comment (yujhe) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: Sounds good, I'll update for that.

**Comment by viirya** [comment]
I think this was fixed recently by #26871.

**Comment by viirya** [comment]
cc @cloud-fan

**Comment by yujhe** [comment]
@viirya I have tested it on the `master` branch, but it still failed on my added test. I think the reason is that the partition key data type might be cast to string type.

**Comment by cloud-fan** [comment]
@yujhe do you have an end-to-end test that other people can run? I'm surprised that the bug is still there.

**Comment by yujhe** [comment]
@cloud-fan here's the test I run in spark-2.4.4., but I have not tested in `master` branch.
```scala
spark.sql("CREATE TABLE timestamp_part (value INT) PARTITIONED BY (dt TIMESTAMP)")
val df = Seq(
    (1, java.sql.Timestamp.valueOf("2019-12-01 00:00:00"), 1), 
    (2, java.sql.Timestamp.valueOf("2019-12-01 01:00:00"), 1)
  ).toDF("id", "dt", "value")
df.write.partitionBy("dt").mode("overwrite").saveAsTable("timestamp_part")

spark.sql("select * from timestamp_part where dt >= '2019-12-01 00:00:00'").explain(true)

```

**Comment by cloud-fan** [comment]
can you test it with master please?

**Comment by yujhe** [comment]
@cloud-fan I have tested in `master` and `branch-2.4` branch with hive-1.2.

`master` branch pass the test, but `branch-2.4` didn't. 
I think the reason is partition key `dt` was not cast to string type in master branch as branch-2.4 did.

We still need to handle the case that partition column may be cast to string or integral type.

Here is the plan executed in master branch.
```
scala> spark.sql("select * from timestamp_part where dt >= '2019-12-01 00:00:00'").explain(true)
== Parsed Logical Plan ==
'Project [*]
+- 'Filter ('dt >= 2019-12-01 00:00:00)
   +- 'UnresolvedRelation [timestamp_part]

== Analyzed Logical Plan ==
id: int, value: int, dt: timestamp
Project [id#19, value#20, dt#21]
+- Filter (dt#21 >= cast(2019-12-01 00:00:00 as timestamp))
   +- SubqueryAlias `default`.`timestamp_part`
      +- Relation[id#19,value#20,dt#21] parquet

== Optimized Logical Plan ==
Filter (isnotnull(dt#21) AND (dt#21 >= 1575129600000000))
+- Relation[id#19,value#20,dt#21] parquet
```

Here's the exception throwed in branch-2.4
```
scala> spark.sql("select * from timestamp_part where dt >= '2019-12-01 00:00:00'").explain(true)                                                                                                            
java.lang.RuntimeException: Caught Hive MetaException attempting to get partition metadata by filter from Hive. You can set the Spark configuration setting spark.sql.hive.manageFilesourcePartitions to fal
se to work around this problem, however this will result in degraded performance. Please report a bug: https://issues.apache.org/jira/browse/SPARK                                                          
  at org.apache.spark.sql.hive.client.Shim_v0_13.getPartitionsByFilter(HiveShim.scala:772)                                                                                                                  
  at org.apache.spark.sql.hive.client.HiveClientImpl$$anonfun$getPartitionsByFilter$1.apply(HiveClientImpl.scala:681)                                                                                       
  at org.apache.spark.sql.hive.client.HiveClientImpl$$anonfun$getPartitionsByFilter$1.apply(HiveClientImpl.scala:679)                                                                                       
  at org.apache.spark.sql.hive.client.HiveClientImpl$$anonfun$withHiveState$1.apply(HiveClientImpl.scala:277)                                                                                               
  at org.apache.spark.sql.hive.client.HiveClientImpl.liftedTree1$1(HiveClientImpl.scala:215)                                                                                                                
  at org.apache.spark.sql.hive.client.HiveClientImpl.retryLocked(HiveClientImpl.scala:214)                                                                                                                  
  at org.apache.spark.sql.hive.client.HiveClientImpl.withHiveState(HiveClientImpl.scala:260)
...
Caused by: java.lang.reflect.InvocationTargetException: org.apache.hadoop.hive.metastore.api.MetaException: Filtering is supported only on partition keys of type string
  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:498)
  at org.apache.spark.sql.hive.client.Shim_v0_13.getPartitionsByFilter(HiveShim.scala:759)
  ... 94 more
Caused by: org.apache.hadoop.hive.metastore.api.MetaException: Filtering is supported only on partition keys of type string
  at org.apache.hadoop.hive.metastore.parser.ExpressionTree$FilterBuilder.setError(ExpressionTree.java:185)
  at org.apache.hadoop.hive.metastore.parser.ExpressionTree$LeafNode.getJdoFilterPushdownParam(ExpressionTree.java:440)
  at org.apache.hadoop.hive.metastore.parser.ExpressionTree$LeafNode.generateJDOFilterOverPartitions(ExpressionTree.java:357)
```

**Comment by cloud-fan** [comment]
Did you use the latest branch-2.4? I'm a little confused about why the bug is still there. No matter we do cast or not, we check the type of the column, not the type of the cast expression.

**Comment by yujhe** [comment]
You are right! The bug is fixed in the latest branch-2.4. 

Here's the result in latest branch-2.4
```
scala> spark.sql("select * from timestamp_part where dt >= '2019-12-01 00:00:00'").explain(true)
== Parsed Logical Plan ==
'Project [*]
+- 'Filter ('dt >= 2019-12-01 00:00:00)
   +- 'UnresolvedRelation `timestamp_part`

== Analyzed Logical Plan ==
id: int, value: int, dt: timestamp
Project [id#16, value#17, dt#18]
+- Filter (cast(dt#18 as string) >= 2019-12-01 00:00:00)
   +- SubqueryAlias `default`.`timestamp_part`
      +- Relation[id#16,value#17,dt#18] parquet

== Optimized Logical Plan ==
Project [id#16, value#17, dt#18]
+- Filter (isnotnull(dt#18) && (cast(dt#18 as string) >= 2019-12-01 00:00:00))
   +- Relation[id#16,value#17,dt#18] parquet
```

**Pull request closed without merging**